### ISSUE TITLE
Petbuild of jwm fails

### ISSUE THREAD

**Issue by peabee** [puppylinux-woof-CE/woof-CE#2873]
Building ScPup32...........
```
Building jwm
checking for gcc... /usr/bin/ccache gcc
checking whether the C compiler works... no
configure: error: in `/tmp/jwm-2.3.7':
configure: error: C compiler cannot create executables
See `config.log' for more details
ERROR: failed to build jwm

# gcc --version
gcc (GCC) 10.3.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

```
Unable to find 'config.log' to see why....
Not sure what to check....

**Comment by dimkr** [comment]
Your compiler is broken, not JWM and not the petbuild.

```
./configure || :
grep -C 10 error config.log
```

**Comment by peabee** [comment]
sandbox3/petbuild-rootfs-complete is broken

/usr/bin/cpp is in sandbox3/devx

but is NOT in sandbox3/petbuild-rootfs-complete

```
# pwd
/mnt/sdc1/builds/sc32/t030222-woof-out_x86_x86_slackware_current/sandbox3/devx/usr/bin
# ls cpp
cpp
# 

# pwd
/mnt/sdc1/builds/sc32/t030222-woof-out_x86_x86_slackware_current/sandbox3/petbuild-rootfs-complete/usr/bin
# ls cpp
ls: cannot access 'cpp': No such file or directory
```
```
checking how to run the C preprocessor... /lib/cpp
configure: error: in `/mnt/sdc1/builds/sc32/t030222-woof-out_x86_x86_slackware_current/sandbox3/ccache-3.7.12':
configure: error: C preprocessor "/lib/cpp" fails sanity check
See `config.log' for more details
make: *** No targets specified and no makefile found.  Stop.
install: cannot stat 'ccache': No such file or directory
install: cannot stat '../petbuild-cache/ccache': No such file or directory

```

**Comment by dimkr** [comment]
Does `chroot sandbox3/devx /lib/cpp --version` work?

**Comment by peabee** [comment]
Yes (devx.sfs loaded):
```
# /lib/cpp --version
cpp (GCC) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

# /usr/bin/cpp --version
cpp (GCC) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

```

**Comment by dimkr** [comment]
```
configure: error: in `/mnt/sdc1/builds/sc32/t030222-woof-out_x86_x86_slackware_current/sandbox3/ccache-3.7.12':
                                                                                                                                                               ^
```

Your problem is elsewhere. You don't have GCC locally, on the host.

Does `cpp --version` work?

**Comment by dimkr** [comment]
btw, it's weird that `gcc --version` says it's version 10, while the preprocessor is from 11. Something about your setup is weird.

**Comment by peabee** [comment]
No - the gcc was the host version - not the target devx version being built.
I do have gcc locally on my building host - but different version to the target system

chroot into sandbox3/devx does not work as its not a complete system - no /lib/ld-linux.so.2

```
# chroot sandbox3/devx lib/cpp --version
chroot: can't execute 'lib/cpp': No such file or directory

# file cpp
cpp: ELF 32-bit LSB executable, Intel 80386, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux.so.2, stripped

```
on test system (not building host) with devx loaded:
```
# /lib/cpp --version
cpp (GCC) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```
Does the building host have to be at same build state as target? This may not always be possible....

**Comment by dimkr** [comment]
> Does the building host have to be at same build state as target?

No.

And `chroot devx` is not supposed to work, literally - you need to chroot into petbuild-rootfs-complete-* for that to work.

**Comment by dimkr** [comment]
Can you share you build configuration so I can build myself?

**Comment by peabee** [comment]
Here is a full transcript of a build attempted on the target system itself with the devx.sfs loaded:
[transcript.txt-false.gz](https://github.com/puppylinux-woof-CE/woof-CE/files/8062449/transcript.txt-false.gz)

Different errors to previous attempts but still fails - /usr/bin/cpp is missing so /lib/cpp is an invalid symlink when chrooted.....

I have also uploaded the petbuild-rootfs-complete as an .sfs to:
https://u.pcloud.link/publink/show?code=XZM8W7VZGbAErIhkYg81Ipe975DCS5RkDabV

**Comment by dimkr** [comment]
Do you still have /mnt/sdc1/builds/sc32/t030222-woof-out_x86_x86_slackware_current/sandbox3/ccache-3.7.12/config.log?

**Comment by peabee** [comment]
[config.log](https://github.com/puppylinux-woof-CE/woof-CE/files/8062510/config.log)

see lines 430/431

**Comment by dimkr** [comment]
Yes, the compiler on the host appears to be broken.

Try this:

```
/bin/echo -e "#include <stdio.h>\n#include <stdlib.h>\n#include <math.h>\n#include <time.h>\nint main() {puts(\"hello\"); free(malloc(1)); return (int)floor((double)time(NULL)/3);}" | gcc -x c -o hello-c - -lm
/bin/echo -e "#include <iostream>\n#include <stdlib.h>\n#include <math.h>\n#include <time.h>\nint main() {std::cout << \"hello\"; free(malloc(1)); return (int)floor((double)time(NULL)/3);}" | g++ -x c++ -o hello-cpp - -lm
```

**Comment by dimkr** [comment]
If you want a quick workaround, add `yes|ccache|ccache|exe>dev,dev` in DISTRO_PKGS_SPECS. petbuilds.sh doesn't try to build a static ccache in the host if the target already has ccache.

**Comment by peabee** [comment]
Would like to get petbuilds to work....

Needs kernel_headers (not in my devx)
JWM still fails to build with /lib/cpp error....
[transcript.log](https://github.com/puppylinux-woof-CE/woof-CE/files/8062922/transcript.log)

from ccache build:
[config.log](https://github.com/puppylinux-woof-CE/woof-CE/files/8062929/config.log)

**Comment by dimkr** [comment]
I'll need JWM's config.log.

Try to modify rootfs-petbuilds/jwm/petbuild:

```
./configure --prefix=/usr --sysconfdir=/etc || :
cat config.log
```

**Comment by peabee** [comment]
[transcript.log](https://github.com/puppylinux-woof-CE/woof-CE/files/8067887/transcript.log)
[pfind-limits.h.log](https://github.com/puppylinux-woof-CE/woof-CE/files/8067889/pfind-limits.h.log)

**Comment by dimkr** [comment]
Your devx (the one you're building) is broken, you'll need to add the Slackware kernel headers package (`kernel-headers`).

**Comment by peabee** [comment]
Still problems.....
sandbox3/devx does have usr/include/X11/Xlib.h but sandbox3/petbuild-rootfs-complete does not

[transcript.log](https://github.com/puppylinux-woof-CE/woof-CE/files/8069649/transcript.log)

**Comment by dimkr** [comment]
You have more missing headers in your devx:

```
conftest.c:38:10: fatal error: X11/X.h: No such file or directory
   38 | #include <X11/X.h>
```

EDIT: specifically, you don't have `xorgproto`

**Comment by peabee** [comment]
Wow that was a struggle - thank you for your help.

**Comment by peabee** [comment]
More help needed if possible..........
Now trying to petbuild JWM on VoidPup..........
[transcript.log](https://github.com/puppylinux-woof-CE/woof-CE/files/8088450/transcript.log)

**Comment by dimkr** [comment]
You don't have msgfmt, you probably don't have gettext in your devx

**Comment by peabee** [comment]
thanks - fixed msgfmt - but build still has problems....
[transcript.log](https://github.com/puppylinux-woof-CE/woof-CE/files/8091108/transcript.log)

**Comment by dimkr** [comment]
It looks like gettext is broken. It doesn't have any output. Do you have all its dependencies?

**Comment by peabee** [comment]
Thanks for your help ............... gettext seems to be OK as far as I can tell........
[transcript.log](https://github.com/puppylinux-woof-CE/woof-CE/files/8096842/transcript.log)

**Comment by peabee** [comment]
Eventually found the gettext problem.......
/packages-templates/gettext_FIXUPHACK
DELETES most of gettext!
`rm -rf usr/share \
	usr/bin/msg* \
	usr/${PLIB}/gettext \
	usr/bin/recode-sr-latin \
	usr/bin/xgettext
`

Fixed by renaming gettext to gettext-void. Still have a problem with the resulting JWM build though - it doesn't support SVG icons - doesn't like the Void librsvg?
`:librsvg:|compat|Packages-void32-current|librsvg-2.52.2_1|librsvg|2.52.2_1|1|Uncategorized|3396K|current|librsvg-2.52.2_1.i686.xbps|+gdk-pixbuf,+glib,+cairo,+libxml2,+pango,+glibc,+libgcc,+librsvg-2.so.2,+libgdk_pixbuf-2.0.so.0,+libgobject-2.0.so.0,+libglib-2.0.so.0,+libcairo-gobject.so.2,+libgio-2.0.so.0,+libxml2.so.2,+libpangocairo-1.0.so.0,+libcairo.so.2,+libpango-1.0.so.0,+libm.so.6,+libdl.so.2,+libgcc_s.so.1,+libc.so.6,+ld-linux.so.2|SVG library for GNOME||||
`

![Screenshot](https://user-images.githubusercontent.com/5926865/154799920-6945adf4-7ccc-4b56-ab3a-02c9dcd13709.png)

**Comment by dimkr** [comment]
Does `pkg-config --cflags --libs librsvg-2.0` work? If not, you probably don't have librsvg, or don't have some of its dependencies. My bet is that you forgot to add librsvg-devel.

**Comment by peabee** [comment]
Seems to be OK:
[transcript.log](https://github.com/puppylinux-woof-CE/woof-CE/files/8102489/transcript.log)
[result-pfilesearch.txt](https://github.com/puppylinux-woof-CE/woof-CE/files/8102491/result-pfilesearch.txt)

librsvg-devel is included.....

**Comment by dimkr** [comment]
Are you checking on the host, or under sandbox3?

The linker says it can't find the library:

```
/usr/bin/ld: cannot find -lrsvg-2.0
```

**Comment by peabee** [comment]
I was checking on the running target (VoidPup32) with the devx installed......

**Comment by dimkr** [comment]
Try `pkg-config --cflags --libs librsvg-2.0`. If it doesn't work, then you don't have librsvg.

**Comment by peabee** [comment]
I did try..........
```
# pkg-config --cflags --libs librsvg-2.0
-pthread -I/usr/include/librsvg-2.0 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng16 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib32/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -lrsvg-2 -lm -lgio-2.0 -lgdk_pixbuf-2.0 -lgobject-2.0 -lglib-2.0 -lcairo

```

**Comment by dimkr** [comment]
Something about your build is messed up. The devx you're building is different from the devx you have loaded on the host, and it's the devx you're building that's used to build JWM.

**Comment by peabee** [comment]
I was testing with "the devx ... that's used to build JWM" - i.e. it was the target devx installed in the target VoidPup32 
Should I upload the .iso and the .devx ???
Is there any way to run the petbuild on the target system with the target devx independently from 3builddistro??

**Comment by peabee** [comment]
To clarify - the messages in transcript.log did come from different tests - the error message comes when doing the petbuild via 3builddistro run on the host; the tests you asked me to do were run on the target system with the target devx installed.

**Comment by dimkr** [comment]
No matter which Puppy you're using to run woof-CE, the devx used to build JWM is the devx of the Puppy you're building. If you want to see if you can build JWM on the host Puppy, just build JWM - you don't need to go through 3builddistro.

**Comment by peabee** [comment]
I understand fully which devx is being used in the petbuild - that is why I was testing the target devx loaded onto the target system.......
I want to see if I can build JWM on the target system with the target devx - not on the host system.....

Do I just directly run:
. ../support/petbuilds.sh
from inside sandbox3? 

Is there any test I can add to petbuilds.sh to show whats happening?
e.g. before line 181 put:
chroot petbuild-rootfs-complete-${NAME} bash -ec "pkg-config --cflags --libs librsvg-2.0"
would this produce any output??

**Comment by dimkr** [comment]
> chroot petbuild-rootfs-complete-${NAME} bash -ec "pkg-config --cflags --libs librsvg-2.0"
> would this produce any output??

Yes, especially if you add `set -x` somewhere before the line you added.

**Comment by peabee** [comment]
Thanks again for your help
Not really sure what the reason is but............. pkg-config is the problem
In Void-Linux it is called pkg-config.pkg-config (600KB) and its not recognised as an executable (tried both symlinking and copying to pkg-config)
By swapping in the Slackware pkg-config (just 40KB) JWM is built correctly with svg support....

e.g. the test line in petbuilds.sh said:
```
Building jwm
++ chroot petbuild-rootfs-complete-jwm bash -ec 'pkg-config --cflags --libs librsvg-2.0'
bash: line 1: pkg-config: command not found
```
https://voidlinux.pkgs.org/current/voidlinux-main-x86_64/pkg-config-0.29.2_3.x86_64.xbps.html shows:
Files
Path
/usr/bin/pkg-config.pkg-config
/usr/share/aclocal/pkg.m4.pkg-config
/usr/share/doc/pkg-config/pkg-config-guide.html
/usr/share/man/man1/pkg-config.1.pkg-config

**Comment by dimkr** [comment]
Why not just `ln -s pkg-config.pkg-config /usr/bin/pkg-config`?

**Comment by peabee** [comment]
I tried that and it didn't work.....

**Comment by peabee** [comment]
Sorry - me again - now struggling with VoidPup64 ........
[transcript.log](https://github.com/puppylinux-woof-CE/woof-CE/files/8110872/transcript.log)

The acl package does have:
/packages-vpup64/acl/usr/lib/libacl.so.1.1.2301 and link libacl.so.1
which is in /sandbox3/rootfs-complete

The devx does have:
/sandbox3/devx/usr/lib/gcc/x86_64-pc-linux-gnu/10.2/liblto_plugin.so.0.0.0 and links
which come from:
yes|gcc_dev|gcc,libgcc,libstdc++,gcc-go,gmp,elfutils,libmpc,isl|exe>dev,dev,doc,nls

So am wondering why the 64-bit petbuild chroot is not finding these items whereas the 32-bit is fine?

A manual run of the jwm petbuild commands in the VoidPup64 target system with the target devx installed runs to completion with no errors..... is there some processing that happens to the devx that is not done for the petbuild chroot?

**Comment by dimkr** [comment]
> So am wondering why the 64-bit petbuild chroot is not finding these items whereas the 32-bit is fine?

Are you using a 32-bit host to build for x86_64? If yes, check if LD_LIBRARY_PATH is set.

**Comment by peabee** [comment]
No - I use a 64-bit host for 64-bit target builds - and a 32-bit host for 32-bit target builds.

**Comment by dimkr** [comment]
Let me know when your woof-CE tree is publicly available and I'll debug it locally.

**Comment by peabee** [comment]
I think the problem relates to the Void 64-bit lib conventions ........ where everything should be in /lib and /usr/lib and /lib64 and /usr/lib64 are symlinks.
This means the devx is malformed if it has a real /lib64 and /usr/lib64 - which it does due to some non-Void components.
I know of no Woof-CE mechanism that could be used to sort this out for both the real devx and the petbuild chroot??
p.s. I suspect the devx sfs-load sorts this out because it refuses to overwrite the lib64 symlinks effectively discarding the problem.

**Comment by dimkr** [comment]
Maybe you have no choice but to build with `USR_SYMLINKS=yes`, and align everything with Void's directory structure. You can't resist this forever, and you'll need to rebuild incompatible packages (so everything is under /usr) if you want a clean solution. 

Some packages look for files in /bin, but they're there only because /bin has the same contents of /usr/bin. You can fight this on a per-file or per-package basis (symlink every file in /bin or /usr/bin to the other directory, or place some hack for problematic packages), but that's horrible and can cause issues in other places (like packages that don't expect a certain directory to be a symlink really).

dpup had the same problems you have, until I went all in on `USR_SYMLINKS=yes`. That makes it incompatible with some old packages (packages with files under /bin, /lib, etc', are a minority), but compatible with a much bigger percentage of Debian packages. Small price to pay and a huge gain, IMHO.

**Comment by peabee** [comment]
Void only symlinks lib64 - not bin and sbin or /usr/bin and /usr/sbin so it is different to both Slackware and Debian....
The main sfs build seems to have no problems (or at least none that can't be fixed with existing Woof-CE mechanisms)
the problem is with the devx and the petbuild chroot which means that petbuilds don't work.
The 32-bit build works because it has no symlinks.

**Comment by dimkr** [comment]
> Void only symlinks lib64

So I guess you'll need to do that too, somehow.

https://voidlinux.org/news/2015/07/usr.sbin-transition.html sounds like a bigger problem, because it can break even packages that don't touch directories outside of /usr.

**Comment by peabee** [comment]
So...what I did as a "proof of concept" was to put a pause-read into 3builddistro at line 369
and then when the build paused I manually merged the contents of devx /usr/lib64 into /usr/lib and then deleted the empty /usr/lib64
With this devx-hack the jwm petbuild completed successfully and so did the overall build.

So the question is - What is the best way to automate this devx-hack ?? I don't think there is an existing Woof-CE mechanism to use?

**Comment by dimkr** [comment]
> So the question is - What is the best way to automate this devx-hack ?? I don't think there is an existing Woof-CE mechanism to use?

The only solution I see is something similar to `USR_SYMLINKS=yes`, with an extra step of "fixing" old .pet packages you rely on, in 2createpackages.

**Comment by peabee** [comment]
Another petbuild problem - this time with UPupJJ+D - I've sent details by PM on the forum. Thanks.

**Comment by peabee** [comment]
Any suggestions / ideas on how to get petbuilds working on UPupJJ+D please? Thanks.

**Comment by dimkr** [comment]
Nope - if I can't build it, I can't debug it.

**Comment by peabee** [comment]
Here is a test-build configuration which builds up to the point that the jwm petbuild fails.....
[ujj-test.tar.gz](https://github.com/puppylinux-woof-CE/woof-CE/files/8392077/ujj-test.tar.gz)

**Comment by dimkr** [comment]
In a quick glance, I have two solutions. Add `-fcf-protection=none` or `-march=i686` to `WOOF_CFLAGS`, under `support/petbuilds.sh`:

https://github.com/puppylinux-woof-CE/woof-CE/blob/587e457021094e9f26e6f6cd54433001e5a71154/woof-code/support/petbuilds.sh#L4

**Comment by peabee** [comment]
Fixed by #3012
